File: week2/14889/14889.py
import sys
input = sys.stdin.readline

class P_14889:
    def __init__(self):
        self.N = int(input())
        self.ability = []
        for _ in range(self.N):
            self.ability.append(list(map(int, input().split())))
        self.total = set(range(self.N))
        self.teamA = []
        self.ans = -1
    
    def run(self):
        self.set_team_and_find_ans(0, 0, self.N/2)
        print(self.ans)

    def set_team_and_find_ans(self, idx, set_num, limit):
        if set_num == limit:
            diff = self.get_team_diff()
            if self.ans == -1:
                self.ans = abs(diff)
            else:
                self.ans = min(self.ans, abs(diff))
            return

        if idx >= self.N or (self.N - idx)+set_num < limit:
            return
        
        # set idx's team to 1
        self.teamA.append(idx)
        self.set_team_and_find_ans(idx+1, set_num+1, limit)
        self.teamA.pop()

        # set idx's team to 0
        self.set_team_and_find_ans(idx+1, set_num, limit)

    def get_team_diff(self):
        teamA_sum = 0
        teamB_sum = 0

        teamB = list(self.total - set(self.teamA))
        for i in range(int(self.N/2)):
            for j in range(i+1, int(self.N/2)):
                teamA_sum += self.ability[self.teamA[i]][self.teamA[j]] + self.ability[self.teamA[j]][self.teamA[i]]
                teamB_sum += self.ability[teamB[i]][teamB[j]] + self.ability[teamB[j]][teamB[i]]
        return teamA_sum - teamB_sum


if __name__=="__main__":
    P = P_14889()
    P.run()

File: with_friends/week2/14889/14889.cpp
#include <iostream>
#include <vector>

using namespace std;

int N;
vector<vector<int> > ability;
vector<int> team;
int ans = -1;

void set_and_find_ans(int, int, int);
int get_team_diff();

int main() {
    cin >> N;
    ability.reserve(N*N);
    team.reserve(N);
    int val;
    for (int i=0; i<N; i++){
        for (int j=0; j<N; j++){
            cin >> val;
            ability[i].push_back(val);
        }
        team[i] = 0;
    }

    set_and_find_ans(0, 0, N/2);
    cout << ans << endl;
}

void set_and_find_ans(int idx, int set_num, int limit) {
    if (set_num == limit) {
        int team_diff = get_team_diff();
        if (ans == -1)
            ans = team_diff;
        else
            ans = min(ans, team_diff);
        return;
    }

    if (idx >= N || (N-idx)+set_num < limit)
        return;

    team[idx] = 1;
    set_and_find_ans(idx+1, set_num+1, limit);
    team[idx] = 0;
    set_and_find_ans(idx+1, set_num, limit);
}

int get_team_diff() {
    int team_sum[2] = {0, 0};
    for (int i=0; i<N-1; i++) {
        for (int j=i+1; j<N; j++) {
            if (team[i] == team[j]) {
                team_sum[team[i]] += ability[i][j] + ability[j][i];
            }
        }
    }
    return abs(team_sum[0] - team_sum[1]);
}